mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -696,3 +702,59 @@ def update_artifacts_location(self, run_id, new_artifacts_location):
         with self.ManagedSessionMaker() as session:
             run = session.query(SqlRun).filter_by(run_uuid=run_id).first()
             run.artifact_uri = new_artifacts_location
+
+
+def _get_attributes_filtering_clauses(parsed):
+    clauses = []
+    for sql_statement in parsed:
+        key_type = sql_statement.get('type')
+        key_name = sql_statement.get('key')
+        value = sql_statement.get('value')
+        comparator = sql_statement.get('comparator')
+        if SearchUtils.is_attribute(key_type, comparator):
+            if comparator in SearchUtils.filter_ops.keys():
+                op = SearchUtils.filter_ops.get(comparator)
+                clauses.append(op(getattr(SqlRun, key_name), value))
+    return clauses
+
+
+def _to_sqlalchemy_filtering_statement(sql_statement, session):
+    key_type = sql_statement.get('type')
+    key_name = sql_statement.get('key')
+    value = sql_statement.get('value')
+    comparator = sql_statement.get('comparator')
+
+    if SearchUtils.is_metric(key_type, comparator):
+        entity = SqlLatestMetric
+        entity_key = SqlLatestMetric.key
+        entity_value = SqlLatestMetric.value
+        value = float(value)
+    elif SearchUtils.is_param(key_type, comparator):
+        entity = SqlParam
+        entity_key = SqlParam.key
+        entity_value = SqlParam.value
+    elif SearchUtils.is_tag(key_type, comparator):
+        entity = SqlTag
+        entity_key = SqlTag.key
+        entity_value = SqlTag.value
+    elif SearchUtils.is_attribute(key_type, comparator):
+        return None
+    else:
+        raise MlflowException("Invalid search expression type '%s'" % key_type,
+                              error_code=INVALID_PARAMETER_VALUE)
+    if comparator in SearchUtils.filter_ops.keys():
+        op = SearchUtils.filter_ops.get(comparator)
+        return session.query(entity).filter(entity_key == key_name, op(entity_value, value)).subquery()
+    else:
+        return None
+
+
+def _get_sqlalchemy_filter_clauses(parsed, session):
+    """creates SqlAlchemy subqueries
+    that will be inner-joined to SQLRun to act as multi-clause filters."""
+    filters = []
+    for sql_statement in parsed:
+        filter_query = _to_sqlalchemy_filtering_statement(sql_statement, session)
+        if filter_query is not None:
```

Comment:
Why not just :

`if filter_query:`

?

Reply:
For some reason it breaks the tests